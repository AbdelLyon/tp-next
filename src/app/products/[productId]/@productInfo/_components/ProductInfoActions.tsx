"use client";
import { Button } from "@/components/ui/button";
import { ShoppingCart, ArrowRight } from "lucide-react";

export const ProductInfoActions = () => {
  return (
    <div className="grid grid-cols-2 gap-4 mt-8" data-testid="product-actions">
      <Button
        size="lg"
        className="font-medium bg-primary hover:opacity-80 shadow-sm hover:shadow transition-all duration-300 rounded-lg"
        data-testid="add-to-cart-button"
        onClick={() => console.log("Add to cart button clicked")}
      >
        <ShoppingCart className="mr-2 h-4 w-4" />
        Add to Cart
      </Button>
      <Button
        size="lg"
        variant="outline"
        className="font-medium border-primary text-primary hover:opacity-80 transition-colors duration-300 rounded-lg"
        data-testid="buy-now-button"
        onClick={() => console.log("Buy now button clicked")}
      >
        Buy Now <ArrowRight className="ml-2 h-4 w-4" />
      </Button>
    </div>
  );
};
