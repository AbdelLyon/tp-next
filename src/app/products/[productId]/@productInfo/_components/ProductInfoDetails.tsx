import { Badge } from "@/components/ui/badge";

interface ProductInfoDetailsProps {
  description: string;
  tags?: string[];
}

export const ProductInfoDetails = ({
  description,
  tags,
}: ProductInfoDetailsProps) => {
  return (
    <>
      <p
        className="text-gray-600 leading-relaxed"
        data-testid="product-description"
      >
        {description}
      </p>

      {tags && tags.length > 0 && (
        <div className="flex flex-wrap gap-2 mt-5" data-testid="product-tags">
          {tags.map((tag) => (
            <Badge
              key={tag}
              variant="secondary"
              className="bg-gray-100 hover:bg-gray-200 text-gray-700 transition-colors"
            >
              {tag}
            </Badge>
          ))}
        </div>
      )}
    </>
  );
};
