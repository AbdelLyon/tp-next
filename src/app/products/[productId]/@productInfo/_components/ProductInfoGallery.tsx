import Image from "next/image";
import { Badge } from "@/components/ui/badge";
import { Button } from "@/components/ui/button";
import { Heart } from "lucide-react";

interface ProductInfoGalleryProps {
  thumbnail?: string;
  title: string;
  discountPercentage?: number;
}

export const ProductInfoGallery = ({
  thumbnail,
  title,
  discountPercentage,
}: ProductInfoGalleryProps) => {
  return (
    <div
      className="relative aspect-square rounded-xl overflow-hidden bg-gray-50 group"
      data-testid="product-gallery"
    >
      {thumbnail && (
        <Image
          src={thumbnail}
          alt={title}
          fill
          className="object-cover transition-transform duration-700 group-hover:scale-105"
          sizes="(max-width: 768px) 100vw, 50vw"
          priority
          unoptimized={true}
        />
      )}

      {discountPercentage && (
        <Badge
          className="absolute top-3 left-3 bg-gradient-to-r from-red-500 to-red-600 text-white border-0 px-3 py-1 text-sm font-medium shadow-sm"
          data-testid="discount-badge"
        >
          {discountPercentage.toFixed(0)}% OFF
        </Badge>
      )}

      <Button
        variant="secondary"
        size="icon"
        className="absolute top-3 right-3 rounded-full opacity-90 hover:opacity-100 border shadow-sm transition-all duration-300 hover:scale-110"
        data-testid="favorite-button"
      >
        <Heart className="h-4 w-4 text-primary" />
      </Button>
    </div>
  );
};
