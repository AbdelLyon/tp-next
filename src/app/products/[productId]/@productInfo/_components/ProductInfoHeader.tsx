import { Badge } from "@/components/ui/badge";
import { Star, Check } from "lucide-react";
import { cn } from "@/lib/utils";

interface ProductInfoHeaderProps {
  brand?: string;
  title: string;
  rating?: number;
  stockLevel: "high" | "medium" | "low" | "out" | null;
  stock?: number;
}

export const ProductInfoHeader = ({
  brand,
  title,
  rating,
  stockLevel,
  stock,
}: ProductInfoHeaderProps) => {
  return (
    <div className="flex flex-col space-y-2 mb-4" data-testid="product-header">
      {brand && (
        <span
          className="text-sm font-medium text-primary/80"
          data-testid="product-brand"
        >
          {brand}
        </span>
      )}

      <h1
        className="text-3xl font-bold tracking-tight bg-clip-text text-transparent bg-gradient-to-r from-gray-800 to-gray-600"
        data-testid="product-title"
      >
        {title}
      </h1>

      <div className="flex items-center gap-2 mt-1">
        {rating && (
          <div className="flex items-center" data-testid="product-rating">
            {[...Array(5)].map((_, i) => {
              const difference = rating - i;
              return (
                <Star
                  key={i}
                  className={cn(
                    "h-4 w-4",
                    difference >= 1
                      ? "text-amber-400 fill-amber-400"
                      : difference >= 0.5
                      ? "text-amber-400 fill-amber-400 opacity-60"
                      : "text-gray-200",
                  )}
                />
              );
            })}
            <span className="text-sm ml-1.5 text-gray-600 font-medium">
              {rating.toFixed(1)}
            </span>
          </div>
        )}

        {stockLevel && (
          <Badge
            variant={stockLevel === "out" ? "destructive" : "outline"}
            className={cn(
              "ml-2",
              stockLevel === "high" &&
                "bg-green-50 text-green-700 border-green-200",
              stockLevel === "medium" &&
                "bg-blue-50 text-blue-700 border-blue-200",
              stockLevel === "low" &&
                "bg-amber-50 text-amber-700 border-amber-200",
            )}
            data-testid="stock-badge"
          >
            {stockLevel === "high" ? (
              <>
                <Check className="inline-block h-3 w-3 mr-1" /> In stock
              </>
            ) : stockLevel === "medium" ? (
              `In Stock (${stock})`
            ) : stockLevel === "low" ? (
              `Low Stock (${stock})`
            ) : (
              "Out of stock"
            )}
          </Badge>
        )}
      </div>
    </div>
  );
};
