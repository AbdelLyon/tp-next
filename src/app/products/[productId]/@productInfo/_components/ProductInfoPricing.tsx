import { Separator } from "@/components/ui/separator";

interface ProductInfoPricingProps {
  price: number;
  originalPrice?: number;
  discountPercentage?: number;
}

export const ProductInfoPricing = ({
  price,
  originalPrice,
  discountPercentage,
}: ProductInfoPricingProps) => {
  return (
    <>
      <div
        className="mt-4 flex items-baseline gap-3"
        data-testid="product-pricing"
      >
        <span className="text-3xl font-bold" data-testid="current-price">
          ${price.toFixed(2)}
        </span>
        {originalPrice && originalPrice > 0 && (
          <span className="text-lg line-through" data-testid="original-price">
            ${originalPrice.toFixed(2)}
          </span>
        )}
        {discountPercentage && originalPrice && (
          <span
            className="text-sm text-green-600 font-semibold bg-content2 px-2 py-0.5 rounded"
            data-testid="savings"
          >
            Save ${(originalPrice - price).toFixed(2)}
          </span>
        )}
      </div>

      <Separator className="my-5 bg-gradient-to-r from-transparent via-gray-200 to-transparent" />
    </>
  );
};
