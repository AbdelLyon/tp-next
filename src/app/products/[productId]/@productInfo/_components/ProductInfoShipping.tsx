import { Truck, Shield, Package } from "lucide-react";

export const ProductInfoShipping = () => {
  return (
    <div
      className="mt-8 space-y-3 bg-gray-50 p-4 rounded-lg border border-gray-100"
      data-testid="shipping-info"
    >
      <div className="flex items-center gap-3 text-sm">
        <div className="h-8 w-8 rounded-full bg-primary/10 flex items-center justify-center text-primary">
          <Truck className="h-4 w-4" />
        </div>
        <span>Free shipping on orders over $50</span>
      </div>
      <div className="flex items-center gap-3 text-sm">
        <div className="h-8 w-8 rounded-full bg-primary/10 flex items-center justify-center text-primary">
          <Shield className="h-4 w-4" />
        </div>
        <span>30-day money-back guarantee</span>
      </div>
      <div className="flex items-center gap-3 text-sm">
        <div className="h-8 w-8 rounded-full bg-primary/10 flex items-center justify-center text-primary">
          <Package className="h-4 w-4" />
        </div>
        <span>Secure packaging and delivery</span>
      </div>
    </div>
  );
};
