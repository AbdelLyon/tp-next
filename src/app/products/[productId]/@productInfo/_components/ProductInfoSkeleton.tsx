import { Skeleton } from "@/components/ui/skeleton";
import { Separator } from "@/components/ui/separator";
import { Tabs, TabsContent, TabsList, TabsTrigger } from "@/components/ui/tabs";

export const ProductInfoSkeleton = () => {
  return (
    <div className="rounded-lg border border-default-200 shadow-sm overflow-hidden">
      <div className="grid grid-cols-1 md:grid-cols-2 gap-8 p-6">
        <div className="relative aspect-square rounded-lg overflow-hidden bg-gray-50 dark:bg-content2">
          <Skeleton className="h-full w-full" />
          <div className="absolute top-3 left-3">
            <Skeleton className="h-6 w-16 rounded-full" />
          </div>
          <div className="absolute top-3 right-3">
            <Skeleton className="h-9 w-9 rounded-full" />
          </div>
        </div>

        <div className="flex flex-col">
          <Skeleton className="h-9 w-4/5 mb-2" />
          <Skeleton className="h-5 w-1/3 mb-4" />

          <div className="flex items-center gap-2 mt-3">
            <Skeleton className="h-4 w-28" />
            <Skeleton className="h-6 w-24 rounded-full ml-2" />
          </div>

          <div className="mt-6 flex items-baseline gap-2">
            <Skeleton className="h-8 w-24" />
            <Skeleton className="h-6 w-20" />
          </div>

          <div className="mt-6 space-y-2">
            <Skeleton className="h-4 w-full" />
            <Skeleton className="h-4 w-full" />
            <Skeleton className="h-4 w-3/4" />
          </div>

          <div className="flex flex-wrap gap-2 mt-6">
            <Skeleton className="h-6 w-16 rounded-full" />
            <Skeleton className="h-6 w-20 rounded-full" />
            <Skeleton className="h-6 w-14 rounded-full" />
          </div>

          <div className="grid grid-cols-2 gap-3 mt-8">
            <Skeleton className="h-11 w-full rounded-md" />
            <Skeleton className="h-11 w-full rounded-md" />
          </div>

          <div className="mt-8 space-y-3">
            <div className="flex items-center gap-2">
              <Skeleton className="h-4 w-4" />
              <Skeleton className="h-4 w-48" />
            </div>
            <div className="flex items-center gap-2">
              <Skeleton className="h-4 w-4" />
              <Skeleton className="h-4 w-56" />
            </div>
          </div>
        </div>
      </div>

      <Separator />

      <div className="p-6">
        <Tabs defaultValue="details">
          <TabsList className="grid w-full grid-cols-3">
            <TabsTrigger value="details">Details</TabsTrigger>
            <TabsTrigger value="specifications">Specifications</TabsTrigger>
            <TabsTrigger value="reviews">Reviews</TabsTrigger>
          </TabsList>
          <TabsContent value="details" className="pt-4">
            <Skeleton className="h-6 w-40 mb-3" />
            <div className="space-y-2">
              <Skeleton className="h-4 w-full" />
              <Skeleton className="h-4 w-full" />
              <Skeleton className="h-4 w-3/4" />
            </div>
          </TabsContent>
        </Tabs>
      </div>
    </div>
  );
};
