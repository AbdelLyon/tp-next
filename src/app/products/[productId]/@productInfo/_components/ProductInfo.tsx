import { productService } from "@/services/ProductService";
import { ProductInfoGallery } from "./ProductInfoGallery";
import { ProductInfoHeader } from "./ProductInfoHeader";
import { ProductInfoPricing } from "./ProductInfoPricing";
import { ProductInfoDetails } from "./ProductInfoDetails";
import { ProductInfoActions } from "./ProductInfoActions";
import { ProductInfoShipping } from "./ProductInfoShipping";
import { ProductInfoTabs } from "./ProductInfoTabs";

const ProductInfo = async ({ productId }: { productId: string }) => {
  const product = await productService.getProductById(productId);

  const originalPrice = product.discountPercentage
    ? product.price * (1 + product.discountPercentage / 100)
    : 0;

  const stockLevel =
    product.stock !== undefined
      ? product.stock > 20
        ? "high"
        : product.stock > 5
        ? "medium"
        : product.stock > 0
        ? "low"
        : "out"
      : null;

  return (
    <>
      <div className="grid grid-cols-1 md:grid-cols-2 gap-8 p-6">
        <ProductInfoGallery
          thumbnail={product.thumbnail}
          title={product.title}
          discountPercentage={product.discountPercentage}
        />

        <div className="flex flex-col">
          <ProductInfoHeader
            brand={product.brand}
            title={product.title}
            rating={product.rating}
            stockLevel={stockLevel}
            stock={product.stock}
          />

          <ProductInfoPricing
            price={product.price}
            originalPrice={originalPrice}
            discountPercentage={product.discountPercentage}
          />

          <ProductInfoDetails
            description={product.description}
            tags={product.tags}
          />

          <ProductInfoActions />

          <ProductInfoShipping />
        </div>
      </div>

      <div className="px-6 pb-6">
        <ProductInfoTabs product={product} />
      </div>
    </>
  );
};

export default ProductInfo;
