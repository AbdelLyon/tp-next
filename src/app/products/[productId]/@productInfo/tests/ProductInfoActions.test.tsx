import { render, screen } from "@testing-library/react";
import { describe, expect, it } from "vitest";
import { RelatedProductsHeader } from "../../@relatedProducts/_components/RelatedProductsHeader";

describe("<RelatedProductsHeader />", () => {
  // Premier test: vérifie que le composant se rend correctement avec le titre
  it("should render the component with title", () => {
    // Rendu du composant dans le DOM virtuel pour les tests
    render(<RelatedProductsHeader title="Related Products" />);

    // Vérifie que l'élément conteneur principal est présent
    expect(screen.getByTestId("section-header")).toBeInTheDocument();

    // Vérifie que le titre est correctement affiché
    expect(screen.getByTestId("section-title")).toHaveTextContent(
      "Related Products",
    );
  });

  // Test pour vérifier que le lien d'action n'est pas rendu sans les props nécessaires
  it("should not render action link when actionLink and actionText are not provided", () => {
    // Rendu du composant avec seulement le titre
    render(<RelatedProductsHeader title="Related Products" />);

    // Vérifie qu'aucun lien d'action n'est présent
    expect(screen.queryByTestId("section-action-link")).not.toBeInTheDocument();
  });

  // Test pour vérifier que le lien d'action est rendu avec les bonnes props
  it("should render action link when actionLink and actionText are provided", () => {
    // Rendu du composant avec titre, lien et texte d'action
    render(
      <RelatedProductsHeader
        title="Related Products"
        actionLink="/products"
        actionText="View All"
      />,
    );

    // Récupération du lien d'action
    const actionLink = screen.getByTestId("section-action-link");

    // Vérifie que le lien est présent
    expect(actionLink).toBeInTheDocument();

    // Vérifie que le texte du lien est correct
    expect(actionLink).toHaveTextContent("View All");

    // Vérifie que l'attribut href est correct
    expect(actionLink).toHaveAttribute("href", "/products");
  });

  // Test pour vérifier que l'icône est correctement rendue
  it("should render icon when icon is provided", () => {
    // Rendu du composant avec une icône de test
    render(
      <RelatedProductsHeader
        title="Related Products"
        actionLink="/products"
        actionText="View All"
        icon={<span data-testid="test-icon">Icon</span>}
      />,
    );

    // Vérifie que l'icône est présente
    expect(screen.getByTestId("test-icon")).toBeInTheDocument();
  });

  // Test pour vérifier que le lien n'est pas rendu avec seulement actionLink
  it("should not render action link when only actionLink is provided", () => {
    // Rendu du composant avec seulement actionLink
    render(
      <RelatedProductsHeader title="Related Products" actionLink="/products" />,
    );

    // Vérifie qu'aucun lien d'action n'est présent
    expect(screen.queryByTestId("section-action-link")).not.toBeInTheDocument();
  });

  // Test pour vérifier que le lien n'est pas rendu avec seulement actionText
  it("should not render action link when only actionText is provided", () => {
    // Rendu du composant avec seulement actionText
    render(
      <RelatedProductsHeader title="Related Products" actionText="View All" />,
    );

    // Vérifie qu'aucun lien d'action n'est présent
    expect(screen.queryByTestId("section-action-link")).not.toBeInTheDocument();
  });

  // Test pour vérifier que les classes CSS sont correctement appliquées
  it("should have correct CSS classes", () => {
    // Rendu du composant
    render(
      <RelatedProductsHeader
        title="Related Products"
        actionLink="/products"
        actionText="View All"
      />,
    );

    // Récupération des éléments
    const sectionHeader = screen.getByTestId("section-header");
    const sectionTitle = screen.getByTestId("section-title");
    const actionLink = screen.getByTestId("section-action-link");

    // Vérifie les classes du conteneur principal
    expect(sectionHeader).toHaveClass("flex items-center justify-between mb-5");

    // Vérifie les classes du titre
    expect(sectionTitle).toHaveClass(
      "text-xl font-semibold bg-gradient-to-r from-primary to-primary/80 bg-clip-text text-transparent",
    );

    // Vérifie les classes du lien d'action
    expect(actionLink).toHaveClass(
      "text-sm text-primary/80 hover:text-primary flex items-center gap-1 transition-colors",
    );
  });

  // Test pour vérifier que le composant gère correctement un titre long
  it("should handle long title correctly", () => {
    // Création d'un titre très long
    const longTitle =
      "This is a very long title that should still render correctly".repeat(3);

    // Rendu du composant avec un titre long
    render(<RelatedProductsHeader title={longTitle} />);

    // Vérifie que le titre long est correctement affiché
    expect(screen.getByTestId("section-title")).toHaveTextContent(longTitle);
  });

  // Test pour vérifier que le lien n'est pas rendu avec une chaîne vide pour actionLink
  it("should not render action link when actionLink is empty string", () => {
    // Rendu du composant avec actionLink vide
    render(
      <RelatedProductsHeader
        title="Related Products"
        actionLink=""
        actionText="View All"
      />,
    );

    // Vérifie qu'aucun lien d'action n'est présent
    expect(screen.queryByTestId("section-action-link")).not.toBeInTheDocument();
  });

  // Test pour vérifier que le lien n'est pas rendu avec une chaîne vide pour actionText
  it("should not render action link when actionText is empty string", () => {
    // Rendu du composant avec actionText vide
    render(
      <RelatedProductsHeader
        title="Related Products"
        actionLink="/products"
        actionText=""
      />,
    );

    // Vérifie qu'aucun lien d'action n'est présent
    expect(screen.queryByTestId("section-action-link")).not.toBeInTheDocument();
  });

  it("should match snapshot", () => {
    // Rendre le composant avec différentes configurations
    const { container: basicComponent } = render(
      <RelatedProductsHeader title="Related Products" />,
    );

    const { container: withActionLink } = render(
      <RelatedProductsHeader
        title="Related Products"
        actionLink="/products"
        actionText="View All"
        icon={<span data-testid="test-icon">Icon</span>}
      />,
    );

    // Vérifier que le rendu correspond aux snapshots enregistrés
    expect(basicComponent).toMatchSnapshot();
    expect(withActionLink).toMatchSnapshot();
  });
});
