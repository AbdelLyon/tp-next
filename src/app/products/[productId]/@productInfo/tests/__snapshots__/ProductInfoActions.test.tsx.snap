// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<RelatedProductsHeader /> > should match snapshot 1`] = `
<div>
  <div
    class="flex items-center justify-between mb-5"
    data-testid="section-header"
  >
    <h2
      class="text-xl font-semibold bg-gradient-to-r from-primary to-primary/80 bg-clip-text text-transparent"
      data-testid="section-title"
    >
      Related Products
    </h2>
  </div>
</div>
`;

exports[`<RelatedProductsHeader /> > should match snapshot 2`] = `
<div>
  <div
    class="flex items-center justify-between mb-5"
    data-testid="section-header"
  >
    <h2
      class="text-xl font-semibold bg-gradient-to-r from-primary to-primary/80 bg-clip-text text-transparent"
      data-testid="section-title"
    >
      Related Products
    </h2>
    <a
      class="text-sm text-primary/80 hover:text-primary flex items-center gap-1 transition-colors"
      data-testid="section-action-link"
      href="/products"
    >
      View All
       
      <span
        data-testid="test-icon"
      >
        Icon
      </span>
    </a>
  </div>
</div>
`;
